### DISCUSSION TITLE
rfc: add request variables support to route definitions

### DISCUSSION THREAD

**Discussion by wei** [vercel/vercel#5062] [Ideas] [answered]
Can certain variables from the request like `host/hostname`, `url`, `path`, maybe even `headers` be made available in route definitions in `vercel.json`?

Something like:
```
{
  "routes": [
    { "src": "/custom-page", "headers": {"x-request-host": "$VERCEL_HOST","x-request-path": "$VERCEL_PATH"}, "dest": "/index.html" },
    { "src": "/([^\?]+)\??(.*)", "dest": "https://my-old-site.com/$1?$2&source=$VERCEL_HOST"},
    { "src": "/redirect", "status": 308, "headers": { "Location": "https://example.com/?from=$VERCEL_URL" } }
  ]
}
```

or php like syntax `$_SERVER['HTTP_HOST']` can work too. What are you guys' thoughts?

**Comment by styfle** [comment, marked as answer]
Thanks for the suggestion! 

This somewhat achievable by reading the [`host`](https://vercel.com/docs/edge-network/headers) request header in your Serverless Function. If you don't want the host that was accessed but rather the unique deployment url, you can use [`x-vercel-deployment-url`](https://vercel.com/docs/edge-network/headers) instead.

However, we might add support for environment variables in `routes`, `rewrites`, or `redirects` sometime in the future.

**Reply by wei** [reply]
Thanks @styfle !  I'm currently using `node-fetch`, setting the response status, body, and looping through all the headers to set them on the response. However, it throws an error: `Error: Parse Error: Content-Length can't be present with Transfer-Encoding`.

To resolve this, I added some logic to only pass through certain headers. So this is what I have so far:
```javascript
import fetch from 'node-fetch';

const PASSTHROUGH_HEADERS = ['location', 'content-type', 'cache-control', 'date'];

export default async (req, res) => {
  const host = req.headers['x-forwarded-host'] || req.headers['host'];
  const proxyUrl = `https://<domain>/${customUrl}`;

  console.log(`Host is ${host}, Proxying to ${proxyUrl}`)

  const response = await fetch(proxyUrl, {
    redirect: 'manual',
  });

  for (let [name, value] of response.headers.entries()) {
    if (PASSTHROUGH_HEADERS.includes(name.toLowerCase())) {
      res.setHeader(name, value);
    }
  }
  
  res.status(response.status);
  res.send(await response.buffer());
}
```

Do you have any suggestions on improving this? Is there a lightweight package out there that can handle rewrite in less lines? I just want the same behavior as `vercel.json - rewrites` with a custom destination. Thank you so much in advance!

**Reply by styfle** [reply]
You could try https://www.npmjs.com/package/http-proxy
